File: internal/keyboard/effects.go
package keyboard

import (
	"context"
	"math"
	"sync"
	"time"
)

// EffectFunc is a per-frame effect function. frame counts from 0.
type EffectFunc func(ctrl *ITE8295, frame int, opts EffectOpts)

// EffectOpts holds parameters for software effects.
type EffectOpts struct {
	Speed   int
	R, G, B byte
}

// DefaultEffectOpts returns sensible defaults.
func DefaultEffectOpts() EffectOpts {
	return EffectOpts{Speed: 3, R: 255, G: 255, B: 255}
}

// EffectRunner manages a software LED effect running in a goroutine.
type EffectRunner struct {
	ctrl   *ITE8295
	fps    int
	cancel context.CancelFunc
	wg     sync.WaitGroup
}

// NewEffectRunner creates a new runner for the given controller.
func NewEffectRunner(ctrl *ITE8295, fps int) *EffectRunner {
	if fps <= 0 {
		fps = 30
	}
	return &EffectRunner{ctrl: ctrl, fps: fps}
}

// Start begins running the effect in a goroutine.
func (r *EffectRunner) Start(fn EffectFunc, opts EffectOpts) {
	r.Stop()
	ctx, cancel := context.WithCancel(context.Background())
	r.cancel = cancel
	r.wg.Add(1)
	go func() {
		defer r.wg.Done()
		r.runLoop(ctx, fn, opts)
	}()
}

// Stop stops the running effect and waits for cleanup.
func (r *EffectRunner) Stop() {
	if r.cancel != nil {
		r.cancel()
		r.wg.Wait()
		r.cancel = nil
	}
}

func (r *EffectRunner) runLoop(ctx context.Context, fn EffectFunc, opts EffectOpts) {
	interval := time.Duration(float64(time.Second) / float64(r.fps))
	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	frame := 0
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			fn(r.ctrl, frame, opts)
			frame++
		}
	}
}

// hsvToRGB converts HSV (h in [0,1), s, v in [0,1]) to RGB bytes.
func hsvToRGB(h, s, v float64) (byte, byte, byte) {
	h -= math.Floor(h) // wrap to [0, 1)
	i := int(h * 6)
	f := h*6 - float64(i)
	p := v * (1 - s)
	q := v * (1 - f*s)
	t := v * (1 - (1-f)*s)

	var r, g, b float64
	switch i % 6 {
	case 0:
		r, g, b = v, t, p
	case 1:
		r, g, b = q, v, p
	case 2:
		r, g, b = p, v, t
	case 3:
		r, g, b = p, q, v
	case 4:
		r, g, b = t, p, v
	case 5:
		r, g, b = v, p, q
	}
	return byte(r * 255), byte(g * 255), byte(b * 255)
}

// RainbowWave is a rainbow wave effect — hue shifts across columns.
func RainbowWave(ctrl *ITE8295, frame int, opts EffectOpts) {
	for row := 0; row < GridRows; row++ {
		for col := 0; col < GridCols; col++ {
			hue := float64(col)/float64(GridCols) + float64(frame)*float64(opts.Speed)*0.005
			r, g, b := hsvToRGB(hue, 1.0, 1.0)
			_ = ctrl.SetKeyColor(row, col, r, g, b)
		}
	}
}

// Breathing is a pulsing brightness effect.
func Breathing(ctrl *ITE8295, frame int, opts EffectOpts) {
	t := float64(frame) * float64(opts.Speed) * 0.02
	factor := (math.Sin(t) + 1.0) / 2.0
	cr := byte(float64(opts.R) * factor)
	cg := byte(float64(opts.G) * factor)
	cb := byte(float64(opts.B) * factor)
	_ = ctrl.SetAllKeys(cr, cg, cb)
}

// ColorWave is a brightness wave that moves across columns.
func ColorWave(ctrl *ITE8295, frame int, opts EffectOpts) {
	for row := 0; row < GridRows; row++ {
		for col := 0; col < GridCols; col++ {
			t := float64(frame) * float64(opts.Speed) * 0.03
			factor := (math.Sin(t-float64(col)*0.5) + 1.0) / 2.0
			r := byte(float64(opts.R) * factor)
			g := byte(float64(opts.G) * factor)
			b := byte(float64(opts.B) * factor)
			_ = ctrl.SetKeyColor(row, col, r, g, b)
		}
	}
}

// SoftwareEffects maps effect names to their functions.
var SoftwareEffects = map[string]EffectFunc{
	"sw_rainbow":   RainbowWave,
	"sw_breathing": Breathing,
	"sw_wave":      ColorWave,
}

File: avellcc/effects.py
"""Software-driven LED effects for the ITE IT8295 keyboard.

These effects run in a Python loop, updating per-key colors each frame.
For hardware-accelerated effects, use ITE8295.set_effect() instead.
"""

import colorsys
import math
import time
import threading

from .ite8295 import ITE8295, GRID_ROWS, GRID_COLS


def _hsv_to_rgb(h: float, s: float = 1.0, v: float = 1.0) -> tuple[int, int, int]:
    r, g, b = colorsys.hsv_to_rgb(h % 1.0, s, v)
    return int(r * 255), int(g * 255), int(b * 255)


class EffectRunner:
    """Runs software-driven LED effects in a background thread."""

    def __init__(self, controller: ITE8295, fps: int = 30):
        self._ctrl = controller
        self._fps = fps
        self._running = False
        self._thread: threading.Thread | None = None

    def start(self, effect_func, **kwargs):
        """Start an effect function in a background thread.

        effect_func(controller, frame, **kwargs) is called each frame.
        It should set key colors using controller.set_key_color().
        """
        self.stop()
        self._running = True
        self._thread = threading.Thread(
            target=self._run_loop,
            args=(effect_func, kwargs),
            daemon=True,
        )
        self._thread.start()

    def stop(self):
        self._running = False
        if self._thread:
            self._thread.join(timeout=2.0)
            self._thread = None

    def _run_loop(self, effect_func, kwargs):
        frame = 0
        interval = 1.0 / self._fps
        while self._running:
            t0 = time.monotonic()
            effect_func(self._ctrl, frame, **kwargs)
            frame += 1
            elapsed = time.monotonic() - t0
            if elapsed < interval:
                time.sleep(interval - elapsed)


def rainbow_wave(ctrl: ITE8295, frame: int, speed: int = 3, **_):
    """Rainbow wave effect - hue shifts across columns."""
    for row in range(GRID_ROWS):
        for col in range(GRID_COLS):
            hue = (col / GRID_COLS + frame * speed * 0.005) % 1.0
            r, g, b = _hsv_to_rgb(hue)
            ctrl.set_key_color(row, col, r, g, b)


def breathing(ctrl: ITE8295, frame: int, r: int = 255, g: int = 255, b: int = 255, speed: int = 3, **_):
    """Breathing effect - pulsing brightness."""
    t = frame * speed * 0.02
    factor = (math.sin(t) + 1.0) / 2.0
    cr = int(r * factor)
    cg = int(g * factor)
    cb = int(b * factor)
    ctrl.set_all_keys(cr, cg, cb)


def color_wave(ctrl: ITE8295, frame: int, r: int = 0, g: int = 100, b: int = 255, speed: int = 3, **_):
    """Color wave - brightness wave moves across columns."""
    for row in range(GRID_ROWS):
        for col in range(GRID_COLS):
            t = frame * speed * 0.03
            factor = (math.sin(t - col * 0.5) + 1.0) / 2.0
            ctrl.set_key_color(row, col, int(r * factor), int(g * factor), int(b * factor))


SOFTWARE_EFFECTS = {
    'sw_rainbow': rainbow_wave,
    'sw_breathing': breathing,
    'sw_wave': color_wave,
}
